go/gossip/storage.go

```diff
@@ -0,0 +1,339 @@
+	insertSCT                     *sql.Stmt
+	insertSCTFeedback             *sql.Stmt
+	insertSTHPollination          *sql.Stmt
+	selectChainID                 *sql.Stmt
+	selectRandomRecentPollination *sql.Stmt
+	selectSCTID                   *sql.Stmt
+
+	selectNumChains   *sql.Stmt
+	selectNumFeedback *sql.Stmt
+	selectNumSCTs     *sql.Stmt
+	selectNumSTHs     *sql.Stmt
+
+	selectFeedback *sql.Stmt
+	selectSTH      *sql.Stmt
+}
+
+// Open opens the underlying persistent data store.
+// Should be called before attempting to use any of the store or search methods.
+func (s *Storage) Open() error {
+	var err error
+	if s.db != nil {
+		return errors.New("attempting to call Open() on an already Open()'d Storage")
+	}
+	s.db, err = sql.Open("sqlite3", *dbName)
+	if err != nil {
+		return err
+	}
+	if _, err := s.db.Exec(schema); err != nil {
+		return err
+	}
+	if s.insertChain, err = s.db.Prepare(insertChain); err != nil {
+		return err
+	}
+	if s.insertSCT, err = s.db.Prepare(insertSCT); err != nil {
+		return err
+	}
+	if s.insertSCTFeedback, err = s.db.Prepare(insertSCTFeedback); err != nil {
+		return err
+	}
+	if s.insertSTHPollination, err = s.db.Prepare(insertSTHPollination); err != nil {
+		return err
+	}
+	if s.selectChainID, err = s.db.Prepare(selectChainID); err != nil {
+		return err
+	}
+	if s.selectRandomRecentPollination, err = s.db.Prepare(selectRandomRecentPollination); err != nil {
+		return err
+	}
+	if s.selectSCTID, err = s.db.Prepare(selectSCTID); err != nil {
+		return err
+	}
+	if s.selectNumChains, err = s.db.Prepare(selectNumChains); err != nil {
+		return err
+	}
+	if s.selectNumFeedback, err = s.db.Prepare(selectNumFeedback); err != nil {
+		return err
+	}
+	if s.selectNumSCTs, err = s.db.Prepare(selectNumSCTs); err != nil {
+		return err
+	}
+	if s.selectNumSTHs, err = s.db.Prepare(selectNumSTHs); err != nil {
+		return err
+	}
+	if s.selectFeedback, err = s.db.Prepare(selectFeedback); err != nil {
+		return err
+	}
+	if s.selectSTH, err = s.db.Prepare(selectSTH); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close closes the underlying DB storage.
+func (s *Storage) Close() error {
+	return s.db.Close()
+}
+
+func selectThingID(getID *sql.Stmt, thing interface{}) (int64, error) {
+	rows, err := getID.Query(thing)
+	if err != nil {
+		return -1, err
+	}
+	if !rows.Next() {
+		return -1, fmt.Errorf("couldn't look up ID for %v", thing)
+	}
+	var id int64
+	if err = rows.Scan(&id); err != nil {
+		return -1, err
+	}
+	return id, nil
+}
+
+// insertThingOrSelectID will attempt to execute the insert Statement (under transaction tx), if that fails due to
+// a unique primary key constraint, it will look up that primary key by executing the getID Statement.
+// Returns the ID associated with persistent thing, or an error describing the failure.
+func insertThingOrSelectID(tx *sql.Tx, insert *sql.Stmt, getID *sql.Stmt, thing interface{}) (int64, error) {
+	txInsert := tx.Stmt(insert)
+	txGetID := tx.Stmt(getID)
```

Comment:
O.M.G.

You should use txInsert.LastInsertId().

Don't believe it needs to be in a transaction, either.